azurelinuxagent/common/logger.py

```diff
@@ -219,18 +226,60 @@ def _should_rollover(self, msg):
         return False
 
     def _do_rollover(self):
+        gz_extension = "gz"
+
         if self._backup_count > 0:
-            for i in range(self._backup_count - 1, 0, -1):
-                src = "%s.%d" % (self.path, i)
-                dst = "%s.%d" % (self.path, i + 1)
-                if os.path.exists(src):
-                    if os.path.exists(dst):
-                        os.remove(dst)
-                    os.rename(src, dst)
-            dst = "%s.1" % self.path
-            if os.path.exists(dst):
-                os.remove(dst)
-            os.rename(self.path, dst)
+            if self._should_archive_backup_files:
+                # Rotating the backup files to make space for the current log file.
+                for i in range(self._backup_count - 1, 0, -1):
+                    src = "%s.%d.%s" % (self.path, i, gz_extension)
+                    dst = "%s.%d.%s" % (self.path, i + 1, gz_extension)
+
+                    # Moving n -> n+1
+                    if os.path.exists(src):
+                        if os.path.exists(dst):
+                            os.remove(dst)
+                        os.rename(src, dst)
+
+                # Backing up the current logging file to logfile.1
+                dst = "%s.1.%s" % (self.path, gz_extension)
+
+                # Removing the logfile.1 that exists already. Already moved above, if needed.
+                if os.path.exists(dst):
+                    os.remove(dst)
+
+                # Archive the current log file. Surrounding by try:except to safeguard against gzip exceptions, if any.
+                try:
+                    with contextlib.closing(gzip.open(dst, 'wb')) as archive:
+                        with open(self.path, 'rb') as logfile:
+                            archive.writelines(logfile)
+                except Exception:
+                    pass  # Didn't add warning here as we are in the process of rotating logs while logging, and
+                    # we don't want to add burden
+
+                # Removing the plain log file as already archived.
+                os.remove(self.path)
+            else:
+                # Rotating the backup files to make space for the current log file.
+                for i in range(self._backup_count - 1, 0, -1):
```

Comment:
Line 234-249 are exactly the same as Line 263-279. Could you please make it a private function instead? Thanks!